### PR TITLE
Fix logical error in vertical merge + replacing merge tree + optimize cleanup

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug which could lead to LOGICAL_ERROR for ReplacingMergeTree + `OPTIMIZE CLEANUP` query.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 4368c00052717b9ad6c69bb6e39a4036621bc0e9 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54368/4368c00052717b9ad6c69bb6e39a4036621bc0e9/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: tests/queries/0_stateless/02861_replacing_merge_tree_with_block_number_column.reference [hunk 1]
```diff
@@ -0,0 +1,13 @@
+== Only last version remains after OPTIMIZE W/ CLEANUP ==
```
Comment (SmitaRKulkarni) [inline tests/queries/0_stateless/02861_replacing_merge_tree_with_block_number_column.reference:1]: Test file name should be changed. 
maybe `xxx_replacing_merge_tree_with_vertical_merge_and_cleanup` ?

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

File: tests/queries/0_stateless/02861_replacing_merge_tree_with_block_number_column.reference (hunk 1 above)
Comment (alesapin) [reply tests/queries/0_stateless/02861_replacing_merge_tree_with_block_number_column.reference:1]: Right

**Review by SmitaRKulkarni — APPROVED** [review]

**Pull request merged**